File: go/core/auto_task.go
package core

import (
	"context"
	"log"
	"time"
)

// NeedsAutoTask returns true if no explicit task is present in the context.
// This is used by adapters to decide whether to create an auto-task from
// ambient ContextData.
func NeedsAutoTask(ctx context.Context) bool {
	return GetCurrentTask(ctx) == nil
}

// CreateAutoTask creates a Task populated from ambient ContextData in the context.
// If no ContextData is set, the task is created with empty attribution fields.
// The taskType should describe the operation being tracked (e.g. "openai.chat",
// "http_request").
func CreateAutoTask(ctx context.Context, taskType string) Task {
	task := NewTask(taskType)
	cd := GetContextData(ctx)
	if cd != nil {
		task.CustomerID = cd.CustomerID
		task.ProjectID = cd.ProjectID
		if cd.Metadata != nil {
			task.Metadata = cd.Metadata
		}
	}
	return task
}

// FinalizeAutoTask completes an auto-created task, aggregates the event
// cost into the task totals, and persists the result to buffer. If task
// is nil the call is a no-op.
func FinalizeAutoTask(task *Task, event *Event, status string, buffer Buffer) {
	if task == nil {
		return
	}

	task.Status = TaskStatus(status)
	now := time.Now().UTC()
	task.EndedAt = &now

	if event != nil {
		switch event.EventType {
		case EventTypeLLMCall:
			task.LLMCostUSD = task.LLMCostUSD.Add(event.CostUSD)
		case EventTypeExternalCost:
			task.ExternalCostUSD = task.ExternalCostUSD.Add(event.CostUSD)
		case EventTypeComputeCost:
			task.ComputeCostUSD = task.ComputeCostUSD.Add(event.CostUSD)
		}
		task.TotalCostUSD = task.LLMCostUSD.Add(task.ExternalCostUSD).Add(task.ComputeCostUSD)

		if event.InputTokens != nil {
			task.TotalInputTokens += *event.InputTokens
		}
		if event.OutputTokens != nil {
			task.TotalOutputTokens += *event.OutputTokens
		}
		if event.CachedTokens != nil {
			task.TotalCachedTokens += *event.CachedTokens
		}
		if event.IsRetry {
			task.RetryCount++
			task.RetryCostUSD = task.RetryCostUSD.Add(event.CostUSD)
		}
	}

	if task.Status == TaskStatusFailed {
		task.FailureCount = 1
	}

	if buffer != nil {
		if err := buffer.UpdateTask(*task); err != nil {
			log.Printf("[dexcost] failed to update auto-task: %v", err)
		}
	}
}

File: typescript/src/core/auto-task.ts
/**
 * Auto-task creation for auto-instrumented calls without an explicit task.
 *
 * Mirrors the Python SDK's `auto_task.py`. When an LLM/HTTP call happens
 * outside an explicit `track()` task, an auto-task is created so the cost
 * is never silently lost. Attribution (customerId/projectId/metadata) is
 * pulled from the ambient `DexcostContext` when available.
 */

import { randomUUID } from "node:crypto";
import { createTask } from "./models.js";
import type { Task } from "./models.js";
import { getCurrentTask, getContext } from "./context.js";
import {
  NetworkAccountant,
  registerAccountant,
} from "../adapters/network-accountant.js";
import { finalizeTaskNetwork } from "./network-finalize.js";
import type { EventBuffer } from "../transport/buffer.js";

/** Return true if there is no active explicit task. */
export function needsAutoTask(): boolean {
  return getCurrentTask() === undefined;
}

/**
 * Create an auto-task with attribution from the current DexcostContext.
 *
 * Reads `customerId`, `projectId`, `metadata`, and `agent` from the
 * ambient context (set via `setContext`). When `agent` is set in the
 * context it overrides the provided `taskType`.
 *
 * The returned task is NOT bound to AsyncLocalStorage. Callers must
 * either pass it explicitly or scope it with `runWithTask()`. The
 * previous design used `setCurrentTask(task)` (enterWith), which
 * leaked the completed task into subsequent calls in the same async
 * chain — a second unwrapped LLM call would inherit the stale task
 * instead of creating its own auto-task.
 */
export function createAutoTask(taskType: string): Task {
  const ctx = getContext();
  const effectiveTaskType = ctx?.agent ? ctx.agent : taskType;
  const task = createTask({
    taskId: randomUUID(),
    taskType: effectiveTaskType,
    customerId: ctx?.customerId,
    projectId: ctx?.projectId,
    metadata: ctx?.metadata ? { ...ctx.metadata } : {},
  });
  // Register a NetworkAccountant so the patched fetch attributes the
  // call's bytes to this task (same as TrackedTask does for explicit
  // tasks). Drained + priced by finalizeAutoTask — every creator of an
  // auto-task MUST finalize it through that helper or the accountant
  // registry entry leaks.
  registerAccountant(task.taskId, new NetworkAccountant());
  return task;
}

/**
 * Finalize an auto-task: set terminal status, drain its NetworkAccountant
 * into byte aggregates + egress dollars, and persist.
 *
 * Counterpart of Python's `finalize_auto_task`, extended with the network
 * finalize step so auto-tasks get the same egress pricing as explicit
 * `tracker.track()` tasks (whose TrackedTask.end() runs the same shared
 * path).
 *
 * Idempotent enough for the instrument call sites: the accountant drain
 * unregisters on first call, and later calls only re-stamp status.
 */
export function finalizeAutoTask(
  task: Task,
  status: "success" | "failed",
  buffer?: EventBuffer | null,
): void {
  task.status = status;
  task.endedAt = new Date();
  if (status === "failed") {
    task.failureCount += 1;
  }
  try {
    finalizeTaskNetwork(task, buffer ?? undefined);
  } catch {
    // Tier-5 fail-silent: a pricing/catalog bug must never break user
    // code paths that finalize auto-tasks (instrument hot paths).
  }
  buffer?.upsertTask(task);
}
